heartbeat: convert to loadable module

garlick opened the pull request:
This PR cleans up the system heartbeat.  It moves heartbeat generation to a loadable module which saves and restores its state in the KVS, so a restarted system instance can continue to generate a monotonic heartbeat sequence number.  The heartbeat protocol is changed, renaming topic string from `hb` to `heartbeat.pulse` and the sequence number from `epoch` to `seq`.

Since the main use case for the heartbeat is synchronizing wake-ups (e.g. for work that is "overhead" from the user's perspective) across the flux instance, a new abstraction for that `flux_sync_create()` is added.  This wraps up the subscribe/unsubscribe and the message handler in a future.  The main motivation is to reuse the future timeout facility to give the synchronized activity an optional "deadline".  If the heartbeat is configured to be infrequent, messages are delayed, or the heartbeat module is unloaded, we can still get our activity done within a specified time bound.  It's also just a little easier to set up than the old way.

The broker overlay "idle peer" detection and the content cache management is switched over to the new abstraction.  Time intervals in those subsystems that were measured in heartbeats are now measured in seconds.  Both will function even if the heartbeat module is not loaded, but if it is loaded, we get the instance-wide synchronization benefit FWIW.

I added a function `flux_future_reset_timeout ()` which can be called after `flux_future_reset()` to reset a timeout, since the only way to do that right now is to call `flux_future_then()` again, which feels like overkill.   A template for setting up and using "sync" with timeouts is thus something like this:
```c
static const double sync_timeout 2.0; // sync on heartbeat, or every timeout (sec), whichever comes first
static flux_future_t *f_sync;

static void sync_cb (flux_future_t *f, void *arg)
{
    if (flux_future_get (f, NULL) < 0 && errno != ETIMEDOUT)
        die ("sync failed");
    do_something ();

    flux_future_reset (f);
    flux_future_reset_timeout (f, sync_timeout);
}

void fini (void)
{
    flux_future_destroy (f_sync);
}

void init (flux_t *h)
{
    if (!(f_sync = flux_sync_create (h))))
        die ("flux_sync_create");
    if (flux_future_then (f_sync, sync_timeout, sync_cb, ctx) < 0)
        die ("flux_future_then");
    ...
}
```

TODO:
* convert kvs to use sync
* unit test for `flux_sync_create()`
* man page for `flux_sync_create()`

grondo commented:
> I added a function flux_future_reset_timeout () which can be called after flux_future_reset() to reset a timeout, since the only way to do that right now is to call flux_future_then() again, which feels like overkill.

I wonder if `flux_future_reset()` should always reset the timeout (and maybe provide an alternate function that doesn't). It seems logical in retrospect that resetting a future  would reset any pending timeout. This might require that timer watchers expose `ev_timer_again()` as something like `flux_timer_watcher_again()` or something, and that the underlying future timeout timer watcher be made repeating.

On a related note, I have often thought it would be nice have a way to provide timers as futures instead of timer watchers, since the `flux_future_t` interface is a bit more convenient. Having the timeout be repeating would allow this to be easily done (though the future would be "fulfilled" with ETIMEDOUT), and to get a repeating timer you could just call `flux_future_reset()` on it, which seems quite natural.

Ok, sorry for the OT rambling...

garlick commented:
> I wonder if `flux_future_reset()` should always reset the timeout (and maybe provide an alternate function that doesn't).

That was my first impulse (and implementation) but I started second-guessing myself, wondering if I was too focused on this one use case.  Thinking about it further, maybe it is the better way to go, and then provide a function that allows the timeout to be changed to another value with -1. as a possibility to disable the timer from recurring.

> This might require that timer watchers expose `ev_timer_again()` as something like `flux_timer_watcher_again()` or similar, and that the underlying future timeout timer watcher be made repeating.

Rereading [libev - be smart about timeouts](http://pod.tst.eu/http://cvs.schmorp.de/libev/ev.pod#code_ev_timer_code_relative_and_opti) it looks like we are using technique 1 which is "almost always a bad idea", so yeah that (technique 2) seems like the way to go.  Let me see if I can puzzle out how to make that work in the future implementation.  Right now we stop the timer when the future is fulfilled.  If we leave it running after non-timeout fulfillment, then we need to protect the future from being fulfilled by the timeout again (queued) if the timer expires before the future if reset.  Since continuations are run by a prep/check, there is a window for that to happen.  Maybe just a flag that says to ignore the timer if it goes off?

garlick commented:
Just pushed a new version of this PR that implicitly resets any timeout in `flux_future_reset()`, adds `flux_timer_watcher_again()` as a wrapper for `ev_timer_again()`, and uses a recurring timer to implement future timeouts instead of starting/stopping a timer.  

I did not add a `flux_future_timer_reset()` interface to allow the timer to be stopped or its value changed b/c I didn't have an immediate use case, but the code is ready for that to be added when it is needed.  It just seemed like a possible waste of time to add it and tests when I we don't have a demonstrated need for it yet.

TODO's are still the same as before.

garlick commented:
Rebased after the various spin-offs.  Now this just converts the built-in broker heartbeat generation to a loadable module.  It's not very exciting.  I think the main argument for it is that with the advent of `flux_sync_create()`, it's now optional, and implemented like this, it's easier to understand than in the broker.  Also the heartbeat period can be altered on the fly by reloading the module now.

grondo commented on src/modules/heartbeat/heartbeat.c:
```diff
@@ -0,0 +1,155 @@
+{
+    struct heartbeat *hb = arg;
+
+    flux_future_destroy (hb->f);
+    if (!(hb->f = flux_event_publish (hb->h, "heartbeat.pulse", 0, NULL)))
+        flux_log_error (hb->h, "error publishing heartbeat");
```
Is `hb->f` really necessary? It seems like it isn't actually used in the module - could you instead use a local `flux_future_t` and destroy it just after creation? Sorry if I missed something obvious.

garlick replied:
That's a good point!  I was sort of half way between that and adding a continuation to capture any publishing errors.  I should probably just add the continuation.

grondo reviewed (approved):
This looks like great cleanup to me!

Just one question inline, but it isn't critical so I've added my approval.

garlick commented:
I'll go ahead and set MWP so I can stir my lunch while the checks run :-)

The future now has a continuation to capture any publishing error.  If a response isn't received before the next heartbeat time, we give up and destroy it (in case the period is set lower than the RTT or there are delays...)

garlick commented:
Except! I just noticed I dropped the broker man page update in the rebase.  I'll tack that on.

grondo commented:
Sounds good! I'll update the NEWS PR.

garlick commented:
OK, force pushed that fix.
